doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

```diff
@@ -0,0 +1,367 @@
+# Graceful Reconfiguration of Managed Clusters
+
+Associated issues/prs:
+[#20010](https://github.com/MaterializeInc/materialize/issues/20010)
+[#26401](https://github.com/MaterializeInc/materialize/pull/26401)
+
+## The Problem
+Reconfiguring a manage dcluster via `alter cluster...` leads to downtime. This
+is due to our cluster sequencer first removing the old replicas then adding new
+replicas that match the new configuration. The duration of downtime can be seen
+as extending through the full period it takes to rehydrate the new replicas.
+
+## Success Criteria
+Provide a mechanism which allows user to delay deletion of old replicas for
+some user defined duraiton while the new replicas spin up. This mechanism should
+allow for a follow on feature, where teardown is delayed until the newly created
+replicas have finished hydrating, or reach some parity with the replicas they
+are replacing.
+
+## V1 Scope
+ - Timeout based reconfiguration cleanup scheduling
+
+## V2 Scope
+ - Hydration based reconfiguration cleanup scheduling
+
+# Out of scope
+ - Issues arising from multi-subscriber message handling.
+
+## Solution Proposal
+
+#### Summary:
+This feature will introduce new SQL in `ALTER CLUSTER`
+which will delay the cleanup of replicas until a specified timeout, or
+until a hydration check is triggered. It will be possible to preempt the cleanup
+by respecifying the statement with a TIMEOUT of 0.
+Ex:
+```sql
+ALTER CLUSTER c1 SET (SIZE 'small') WITH ( CLEANUP [AUTO DETECT] TIMEOUT 5 minutes );
+```
+Here, `AUTO DETECT` indicates we would automatically detect via some parity check that the
+new replicas are as hydrated as the prior replicas. For now a timeout must be specified.
+
+The new syntax and mechanisms will be built to work with auto cleanup mechanisms
+built around cluster hydration but the syntax for that and the mechanism are not
+in scope for this document.
+
+All replicas will be billed during reconfiguraiton (except explicitly unbilled replicas).
+
+
+#### Definitions:
+  - Active Replica: These replicas should be actively serving results and are
+  named r1,r2,r3,etc... if a reconfiguration is ongoing they will match the
+  prior version of the configuration. If a reconfiguration is not ongoing they
+  should be the only replicas.
+  - Reconfiguration Replica: These replicas only exist during a
+  reconfiguration , their names will be suffixed with `reconfig-`. They will
+  move to active once the reconfiguration is finalized.
+
+
+#### Guard Rails
+- Only one reconfigure can occur at a time for a given cluster. If a cluster
```

Comment:
This is exactly what I was thinking. There's a small bit on the cluster sequencing portion on this flow, but I'll make it clear throughout that this will be an ovewrite of the existing reconfigure.